drivers/ethernet/eth_w5500.c

```diff
@@ -0,0 +1,586 @@
+/* W5500 Stand-alone Ethernet Controller with SPI
+ *
+ * Copyright (c) 2020 Linumiz
+ * Author: Parthiban Nallathambi <parthiban@linumiz.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT	wiznet_w5500
+
+#define LOG_MODULE_NAME	eth_w5500
+#define LOG_LEVEL	CONFIG_ETHERNET_LOG_LEVEL
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#include <zephyr.h>
+#include <device.h>
+#include <string.h>
+#include <errno.h>
+#include <drivers/gpio.h>
+#include <drivers/spi.h>
+#include <net/net_pkt.h>
+#include <net/net_if.h>
+#include <net/ethernet.h>
+#include <ethernet/eth_stats.h>
+
+#include "eth.h"
+#include "eth_w5500_priv.h"
+
+#define WIZNET_OUI_B0	0x00
+#define WIZNET_OUI_B1	0x08
+#define WIZNET_OUI_B2	0xdc
+
+#define W5500_SPI_BLOCK_SELECT(addr)	(((addr) >> 16) & 0x1f)
+#define W5500_SPI_READ_CONTROL(addr)	(W5500_SPI_BLOCK_SELECT(addr) << 3)
+#define W5500_SPI_WRITE_CONTROL(addr)   \
+	((W5500_SPI_BLOCK_SELECT(addr) << 3) | BIT(2))
+
+static int w5500_spi_read(struct device *dev, uint32_t addr,
+			  uint8_t *data, uint32_t len)
+{
+	int ret;
+	struct w5500_runtime *ctx = dev->data;
+	/* 3 bytes as 0x010203 during command phase */
+	uint8_t tmp[len + 3];
+
+	uint8_t cmd[3] = {
+		addr >> 8,
+		addr,
+		W5500_SPI_READ_CONTROL(addr)
+	};
+	const struct spi_buf tx_buf = {
+		.buf = cmd,
+		.len = ARRAY_SIZE(cmd),
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1,
+	};
+	const struct spi_buf rx_buf = {
+		.buf = tmp,
+		.len = ARRAY_SIZE(tmp),
+	};
+	const struct spi_buf_set rx = {
+		.buffers = &rx_buf,
+		.count = 1,
+	};
+
+	k_sem_take(&ctx->tx_rx_sem, K_FOREVER);
```

Comment:
this seems useless to me. SPI controller drivers already have an internal resource lock.

Reply:
Thanks, resolved.